monzo/monzo.py

```diff
@@ -276,3 +277,57 @@ def update_transaction_notes(self, transaction_id, notes):
            :rtype: The updated transaction object.
         """
         return self.update_transaction_metadata(transaction_id, 'notes', notes)
+
+    def create_receipt(self, transaction_id, total, currency, items, taxes=None, payments=None, merchant=None):
+        """Create a receipt for a given transaction. (https://docs.monzo.com/#create-receipt) (https://docs.monzo.com/#receipts)
+           :param transaction_id: The unique identifier for the transaction for which the receipt should be added to.
+           :param total: The integer amount for the value of the transaction.
+           :param currency: The str of the currency used for the transaction; usually `GBP`.
+           :param items: The list of items detailing the products included in the total. (https://docs.monzo.com/#receipt-items)
+           :param taxes: Not required by default: The list of taxes (e.g. VAT) added onto the total. (https://docs.monzo.com/#receipt-taxes)
+           :param payments: Not required by default: A list of payments, indicating how the customer paid the total. (https://docs.monzo.com/#receipt-payments)
+           :param merchant: Not required by default: Extra information about the merchant you shopped at. (https://docs.monzo.com/#receipt-merchant)
+           :rtype: A tuple object containing the receipt object (receipt_id and infoformation) you "PUT", repeated back to you & the generate external_id.
+        """
+        url = "{0}/transaction-receipts".format(self.API_URL)
+        # Generate a random external_id consisting of 8 digits. 
+        external_id = "Order-" + str(random.randint(10000000, 99999999))
+        # Init base receipt structure with required properties.
+        data = {
+            "transaction_id": transaction_id,
+            "external_id": external_id,
+            "total": total,
+            "currency": currency,
+            "items": items
+        }
+        # Check for non-required properties and are the correct type (taxes, payments & merchant).
+        if (taxes is not None) and (type(taxes) is list):
```

Comment:
So instead of the checks have something along the lines of:
```python
data = {
            "transaction_id": transaction_id,
            "external_id": external_id,
            "total": total,
            "currency": currency,
            "items": items,
            "taxes": taxes (this would be None if not specified),
            "payments": payments (this would be None if not specified),
            "merchant": merchant (this would be None if not specified)
        }
```
This would work you recon?

Reply:
Yep, although with pregenerated receipts we can remove these lines and directly attach the input to the request.

Reply:
Ok, I think it would be best to go down the route of having one `:param: receipt` which will be the entire structure for the receipt. This structure would already include the `transaction_id` so no need for the param `transaction id`.

So the function would be `create_receipt(self, receipt)`

Sounds good?

Reply:
The external_id can already be supplied within the receipt object. If we were to go with `create_receipt(self, receipt, external_id)` then we would need to add:
```python
def create_receipt(self, receipt, external_id):
        ...
        # Add external_id to the receipt object.
        receipt["external_id"] = external_id
```
Otherwise the parameter would be pointless.

Reply:
Hmm, I guess so. I think its best to be explicit for this parameter given its importance.

I'd change that line to act on a copy of the receipt as you're silently mutating the original dict as it stands.